cmd/kubeadm/app/cmd/upgrade/plan.go

```diff
@@ -155,11 +154,33 @@ func printAvailableUpgrades(upgrades []upgrade.Upgrade, w io.Writer, featureGate
 		fmt.Fprintf(tabw, "Controller Manager\t%s\t%s\n", upgrade.Before.KubeVersion, upgrade.After.KubeVersion)
 		fmt.Fprintf(tabw, "Scheduler\t%s\t%s\n", upgrade.Before.KubeVersion, upgrade.After.KubeVersion)
 		fmt.Fprintf(tabw, "Kube Proxy\t%s\t%s\n", upgrade.Before.KubeVersion, upgrade.After.KubeVersion)
-		if features.Enabled(featureGates, features.CoreDNS) {
-			fmt.Fprintf(tabw, "CoreDNS\t%s\t%s\n", upgrade.Before.DNSVersion, upgrade.After.DNSVersion)
+
+		printCoreDNS, printKubeDNS := false, false
+		coreDNSBeforeVersion, coreDNSAfterVersion, kubeDNSBeforeVersion, kubeDNSAfterVersion := "", "", "", ""
+
+		if upgrade.Before.DNSType == constants.CoreDNS {
+			printCoreDNS = true
+			coreDNSBeforeVersion = upgrade.Before.DNSVersion
+		} else {
+			printKubeDNS = true
+			kubeDNSBeforeVersion = upgrade.Before.DNSVersion
+		}
+
+		if upgrade.After.DNSType == constants.CoreDNS {
+			printCoreDNS = true
+			coreDNSAfterVersion = upgrade.After.DNSVersion
 		} else {
-			fmt.Fprintf(tabw, "Kube DNS\t%s\t%s\n", upgrade.Before.DNSVersion, upgrade.After.DNSVersion)
+			printKubeDNS = true
+			kubeDNSAfterVersion = upgrade.After.DNSVersion
 		}
+
+		if printCoreDNS {
+			fmt.Fprintf(tabw, "CoreDNS\t%s\t%s\n", coreDNSBeforeVersion, coreDNSAfterVersion)
+		}
+		if printKubeDNS {
+			fmt.Fprintf(tabw, "Kube DNS\t%s\t%s\n", kubeDNSBeforeVersion, kubeDNSAfterVersion)
+		}
+
```

Comment:
given the `DNSType` extension i would refactor the above to be a `switch / case` instead of handling it with `if / then`. this would then need less line changes once more DNS types are introduced.

Reply:
Latest updates the conditionals that set printKubeDNS/printCoreDNS and the before/after versions, but since we need to potentially print multiples I think it still made sense to leave the if statements in place for the actual printing of the output.

Reply:
a map can be used too, but it kind of depends if it's OK to print names in the tabs `\t`:
this also solves the case where the before and after types differ (?). to my understanding this was possible.

```php
// define names for known DNS types here
var dnsNames = make(map[string]string)
dnsNames[constants.CoreDNS] = "CoreDNS"
dnsNames[constants.KubeDNS] = "KubeDNS"

dnsBefore, isKnownDNSBefore := dnsNames[upgrade.Before.DNSType]
dnsAfter, isKnownDNSAfter := dnsNames[upgrade.After.DNSType]

if isKnownDNSBefore && isKnownDNSAfter {
	fmt.Printf("DNS upgrade\t%s %s\t%s %s\n", dnsBefore, upgrade.Before.DNSVersion, dnsAfter, upgrade.After.DNSVersion)
} else {
	// do we need to print an error?
	fmt.Printf("DNS upgrade encountered unknown type. Before: %s, After: %s\n", upgrade.Before.DNSType, upgrade.After.DNSType)  
}
```

Reply:
Oooh, I like that!

Reply:
@detiber I'm not sure I got the full problem here, but I think that we should consider also to use mutators to "fix" older configuration in order to adapt to the latest change in code.
In this case, when upgrading, if the fetureflag CoreDNS is not set, we should add fetureflag coreDNS=false in order to avoid an unexpected switch to CoreDNS.
WDYT?

Reply:
@fabriziopandini There is another PR pending that helps improve the migration between dns addons here: https://github.com/kubernetes/kubernetes/pull/63400

I do think that we *should* upgrade to CoreDNS by default and that we need to document this not only in the release notes, but also the other kubeadm documentation so that users that wish to stay on Kube DNS will know that they need to provide the proper command line arguments (or config) to do so.

Reply:
@fabriziopandini If this was just a case of changing a default for a choice of options that were going to be supported long term, then I would agree with trying to use some type of a mutator to persist the existing config.  However, since Kube DNS is being phased out, I think a user should be required to opt-in in order to keep using it.

Reply:
> Ideally, in the case that we are replacing a component we should break it out from the component upgrade output and draw attention to it in a way that makes it clear to the user what is happening.

> @neolit123 do you have any objections with continuing that discussion on a separate issue that can be addressed as a followup?

@detiber sounds good, that would probably be the correct course. :+1:

Reply:
https://github.com/kubernetes/kubeadm/issues/810

Reply:
Put a // TODO: comment with the notes above this and link to a separate issue please.

Reply:
@timothysc done